py/objtuple.c

```diff
@@ -31,6 +31,9 @@
 #include "py/objtuple.h"
 #include "py/runtime.h"
 
+// type check is done on getiter method to allow tuple, namedtuple, attrtuple
+#define mp_obj_is_tuple_compatible(o) (mp_obj_is_obj(o) && ((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type->getiter == mp_obj_tuple_getiter)
```

Comment:
To retain existing behaviour/code size I think this should be exactly what was in `tuple_cmp_helper`, namely `mp_obj_get_type(o)->getiter == mp_obj_tuple_getiter`.

(Sorry, I was very rough before when I suggested to copy how objdict works.)